Handle yaml merge keys correcly.

edge90 opened the pull request:
## Description:


**Related issue (if applicable):** fixes https://github.com/esphome/issues/issues/886

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

edge90 committed:
Handle yaml merge keys correcly.

edge90 committed:
Removed old debug bool.

OttoWinter commented on esphome/yaml_util.py:
```diff
@@ -130,50 +130,45 @@ def construct_yaml_seq(self, node):
-            node.value = node.value + post_merge
+        if merge:
+            def determine(key, value, haystack):
+                return key.value in haystack
```
This line doesn't look right - it will access the attribute .value in `key` (as in the parameter called `value` literally). The parameter `value` to this function is unused.

edge90 replied:
Without having the code infront of me I'm pretty sure that key.value contains the the key name, like platform, pin etc. value.value would contain gpio, GPIO0 etc.

The whole point of that is to not add keys that should be overridden in the next step.

OttoWinter commented on tests/merge_anchors.yaml:
```diff
@@ -0,0 +1,20 @@
+substitutions: &esphome
```
Unfortunately, this file needs to be removed - the problem with the tests folder is that each test file adds a significant amount of time to the CI checks and thus needs to be limited.

You can try to add YAML anchors to the existing test files though.

edge90 replied:
Ok, will do!

edge90 replied:
Removed, misspelled your name though :/

OttoWinter reviewed (commented):
Please also explain how you came up with this code change - it's hard to analyze this change (and check it for correctness) from the diff alone.

Is this based on the updated PyYAML codebase? How does it handle duplicate keys in *normal* dicts (without merge keys)? The new code should definitely still raise an error if there are normal duplicate keys.

edge90 commented:
I was trying to avoid config/code duplication and wanted to solve it by using anchors. The normal python implemention does not complain if you have duplicate keys. This implemention still does if you omit the merge "operator".

And yes, it's based on the PyYAML implemention which I tweaked to handle the merge stuff (as PyYAML doesn't need to do as it doesn't care about duplicate keys)

edge90 committed:
Deleted after request from @OttoWinder.

edge90 committed:
Small refactoring.

Removed unused variable `value`
Small refactoring to make the code clearer.
Added comments.

OttoWinter commented:
Hi, thanks a lot!

I looked over the implementation and the [YAML spec](https://yaml.org/type/merge.html) and found one edge case:

If a merge sequence is used, the YAML spec says the merge entries defined earlier in the sequence should override those later in the sequence. The current code returns a duplicate-key error for this input:
```yaml
.test: &test
  initial_value: 'value1'
.test2: &test2
  initial_value: 'value2'

globals:
  - id: something
    type: 'int'
    <<: [*test, *test2]
```

I tweaked the merge code a bit to fix that edge case (probably not too many people know about that merge sequence feature anyway, but still) and now it is fixed.

I've pushed to code, if you're ok with the changes I'd be happy to merge.

OttoWinter committed:
Fix merge sequence edge case

OttoWinter committed:
Merge branch 'dev' into dev

edge90 commented:
Fantastic, good catch! 

Looks fine to me so you can go ahead and merge :)
